contrib/scalar: add the `version` subcommand

dscho opened the pull request:
This command is handled somewhat magically in the .NET version, and it is a bit unclear to this developer how the version is updated... Maybe we can start with a hard-coded version, though, and make it more dynamic (e.g. generated from the recent dates of the commit history of `contrib/scalar/`) later?

dscho committed:
fixup! Implement `scalar diagnose`

Let's really include the full Git version in the `diagnose` output.

It is unclear why there is a link error when calling
`get_version_info()` and `load_builtin_commands()` is missing (nothing
in `get_version_info()` comes even close to building the command list),
but providing a dummy implementation of the latter function works around
it.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: implement the `version` command

Since Scalar is now tightly coupled with Git, it does not make sense for
them to show different versions.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
One option to generate this version number dynamically would be:

```sh
git -C contrib/scalar log --since=4.weeks.ago --format=%as -- \
  Makefile scalar.c scalar.h json-parser.h json-parser.c |
sort -nr |
cut -c 3-7 |
uniq -c |
sed -n '1s| *\(.*\) \(.*\)-\(.*\)|\2.\3.\1.0|p'
```

This could be generated via an additional `Makefile` rule that writes, say, `contrib/scalar/scalar-version.h`.

derrickstolee commented:
Hm. I was thinking this would just be simple plumbing to the Git version. Do you see an important reason to keep the Scalar version format as-is?

Yes, it will look like a backwards jump, but will that cause a significant problem to anyone?

A question for @ldennington: do you compare Scalar version strings at all in the tooling you know about?

mjcheetham commented:
I think pointing to the Git version should be "good enough". The only places where _Scalar_ version numbers are compared IIRC would be in the old upgrader logic for `scalar upgrade`.

ldennington commented:
> Hm. I was thinking this would just be simple plumbing to the Git version. Do you see an important reason to keep the Scalar version format as-is?
> 
> Yes, it will look like a backwards jump, but will that cause a significant problem to anyone?
> 
> A question for @ldennington: do you compare Scalar version strings at all in the tooling you know about?

Indeed, we do this in osetup. But once this feature lands and my updates to only install microsoft/git and the telemetry tool go in we will no longer do that comparison.

dscho commented:
I guess I am just infatuated with the idea I mentioned in https://github.com/microsoft/git/pull/338#issuecomment-837336588... ;-)

dscho committed:
scalar: implement the `help` subcommand

It is merely handing off to `git help scalar`.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho commented:
I rebased onto the latest version of `features/scalar`. Now, when calling `scalar version` it actually prints the same as `git version` (but for backwards-compatibility, `scalar version` prints to `stderr`, unlike `git version`, which prints to `stdout`).

While at it, I also added the `scalar help` function (which depends on [the hack](https://github.com/microsoft/git/commit/8509e0246762a3957d65bcbd1d02444f44c0c98b) where `git help scalar` will open Scalar's documentation).
